orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/kato/pipeline/support/ResizeStrategySupport.groovy

```diff
@@ -50,4 +66,90 @@ public class ResizeStrategySupport {
+                                             String cloudProvider,
+                                             String location) {
+    TargetServerGroup tsg = oortHelper.getTargetServerGroup(account, serverGroupName, location, cloudProvider)
+      .orElseThrow({
+      new IllegalStateException("no server group found $cloudProvider/$account/$serverGroupName in $location")
+    })
+
+    def currentMin = Integer.parseInt(tsg.capacity.min.toString())
+    def currentDesired = Integer.parseInt(tsg.capacity.desired.toString())
+    def currentMax = Integer.parseInt(tsg.capacity.max.toString())
+
+    return new ResizeStrategy.Capacity(currentMax, currentDesired, currentMin)
+  }
+
+  ResizeStrategy.Capacity pinMin(ResizeStrategy.Capacity capacity) {
+    return new ResizeStrategy.Capacity(
+      capacity.max,
+      capacity.desired,
+      capacity.desired
+    )
+  }
+
+  ResizeStrategy.Capacity unpinMin(ResizeStrategy.Capacity capacity, Integer originalMin) {
+    if (originalMin > capacity.min) {
+      log.warn("Can not unpin the minimum of ${capacity} to a higher originalMin=${originalMin}")
+      return capacity
+    }
+
+    return new ResizeStrategy.Capacity(
+      capacity.max,
+      capacity.desired,
+      originalMin)
+  }
+
+
+  ResizeStrategy.Capacity pinCapacity(ResizeStrategy.Capacity capacity) {
+    return new ResizeStrategy.Capacity(
+      capacity.desired,
+      capacity.desired,
+      capacity.desired
+    )
+  }
+
+  ResizeStrategy.Capacity scalePct(ResizeStrategy.Capacity capacity, double factor) {
+    // scalePct only applies to the desired capacity
+    def newDesired = (Integer) Math.ceil(capacity.desired * factor)
+    return new ResizeStrategy.Capacity(
+      Math.max(capacity.max, newDesired), // in case scalePct pushed desired above current max
+      newDesired,
+      Math.min(capacity.min, newDesired)) // in case scalePct pushed desired below current min
+  }
+
+  ResizeStrategy.Capacity performScalingAndPinning(Capacity sourceCapacity,
+                                                   Stage stage,
+                                                   ResizeStrategy.OptionalConfiguration resizeConfig) {
+    ResizeStrategy.StageData stageData = stage.mapTo(ResizeStrategy.StageData)
+
+    def newCapacity = sourceCapacity
+    if (resizeConfig.scalePct != null) {
+      double factor = resizeConfig.scalePct / 100.0d
+      newCapacity = scalePct(sourceCapacity, factor)
+    }
+
+    if (stageData.unpinMinimumCapacity) {
+      Integer originalMin = null
+      def originalSourceCapacity = stage.context.get("originalCapacity.${stageData.source?.serverGroupName}".toString()) as Capacity
+      originalMin = originalSourceCapacity?.min ?: stage.context.savedCapacity?.min
+
+      if (originalMin != null) {
+        newCapacity = unpinMin(newCapacity, originalMin as Integer)
+      } else {
+        log.warn("Resize stage has unpinMinimumCapacity==true but could not find the original minimum value in stage" +
```

Comment:
would it be worth throwing a counter in here so we could alert on this case happening? ( I don't think there is a normal case where this would happen? )

Reply:
in a7aa14472a990c23c13bbe2d2ecf9f2928af1afb